presto-native-execution/presto_cpp/main/operators/BroadcastExchangeSource.cpp

```diff
@@ -34,26 +34,29 @@ folly::SemiFuture<BroadcastExchangeSource::Response>
 BroadcastExchangeSource::request(
     uint32_t /*maxBytes*/,
     uint32_t /*maxWaitSeconds*/) {
-  std::vector<velox::ContinuePromise> promises;
+  if (atEnd_) {
+    return folly::makeFuture(Response{0, true});
+  }
+
+  bool hasNext = reader_->hasNext();
   int64_t totalBytes = 0;
+  std::unique_ptr<velox::exec::SerializedPage> page;
+  if (hasNext) {
+    // Read outside the lock to avoid a potential deadlock
+    // ExchangeClient guarantees not to call ExchangeSource#request concurrently
+    auto buffer = reader_->next();
+    totalBytes = buffer->size();
+    auto ioBuf = folly::IOBuf::wrapBuffer(buffer->as<char>(), buffer->size());
+    page = std::make_unique<velox::exec::SerializedPage>(
+        std::move(ioBuf), [buffer](auto& /*unused*/) {});
+  }
+
+  std::vector<velox::ContinuePromise> promises;
   {
+    // Limit locking scope to queue manipulation
     std::lock_guard<std::mutex> l(queue_->mutex());
-    if (atEnd_) {
-      return folly::makeFuture(Response{0, true});
-    }
-
-    if (!reader_->hasNext()) {
-      atEnd_ = true;
-      queue_->enqueueLocked(nullptr, promises);
-    } else {
-      auto buffer = reader_->next();
-      totalBytes = buffer->size();
-      auto ioBuf = folly::IOBuf::wrapBuffer(buffer->as<char>(), buffer->size());
-      queue_->enqueueLocked(
-          std::make_unique<velox::exec::SerializedPage>(
-              std::move(ioBuf), [buffer](auto& /*unused*/) {}),
-          promises);
-    }
+    atEnd_ = !hasNext;
+    queue_->enqueueLocked(std::move(page), promises);
   }
   for (auto& promise : promises) {
     promise.setValue();
```

Comment:
Let's take
```
atEnd_ = !hasNext;
```
out of the lock protection.
Otherwise, there is a question: why is it protected here, but not in the beginning of the method.